src/test/com/fulcrologic/fulcro/algorithms/normalized_state_helpers_spec.cljc

```diff
@@ -0,0 +1,222 @@
+    (let [state state]
+      (assertions
+        (nsh/tree-path->db-path state [:b 1 :c])
+        => [:d 1])))
+
+  (behavior "Resolves table-nested to-many references"
+    (let [state state]
+      (assertions
+        (nsh/tree-path->db-path state [:b 1 :e 0])
+        => [:b 1])))
+
+  (behavior "Returns the given tree-path for denormalized data"
+    (let [state state]
+      (assertions
+        (nsh/tree-path->db-path state [:denorm :level-1 :level-2])
+        => [:denorm :level-1 :level-2])))
+
+  (behavior "Returns not-found (if the parameter is provided) when the path doesn't exist"
+    (let [state state]
+      (assertions
+        (nsh/tree-path->db-path state [:root/person 4 :person/id] "not-found")
+        => "not-found"))))
+
+;;============================================================================
+
+(specification "get-in" :focus
+  (behavior "Follows edges from root that are denormalized"
+    (let [state {:a [:b 1]
+                 :b {1 {:c [:d 1]}}
+                 :d {1 {:value 42}}}]
+      (assertions
+        (nsh/get-in state [:a :c :value]) => 42)))
+
+  (behavior "Behaves like a normal clojure get-in"
+    (let [state {:a {:c {:value 42}}}]
+      (assertions
+        (nsh/get-in state [:a :c :value]) => 42)))
+
+  (behavior "Returns nil when the value isn't found"
+    (let [state state]
+      (assertions
+        (nsh/get-in state [:a :b :c]) => nil)))
+
+
+  (behavior "Returns not-found when provided this option"
+    (let [state state]
+      (assertions
+        (nsh/get-in state [:a :b :c] "not-found") => "not-found")))
+
+  )
+
+
+;;============================================================================
+
+(specification "remove-entity*"
+  (let [denorm-data {:a [[:b 1]] :b [:b 1]}
+        state {:a      [:b 1]
+               :top-tm [[:d 1] [:b 1]]
+               :denorm {:level-1 {:level-2 denorm-data}}
+               :b      {1 {:c [:d 1]
+                           :e [[:b 1] [:d 1]]}}
+               :d      {1 {:value 42}}}]
+    (assertions
+      "Removes the entity itself from the database"
+      (-> (nsh/remove-entity* state [:b 1])
+          (get-in [:b 1])
+          nil?) => true
+      "Removes top-level to-one references"
+      (-> (nsh/remove-entity* state [:b 1]) :a nil?) => true
+      "Removes table-nested to-one references"
+      (-> (nsh/remove-entity* state [:d 1]) (get-in [:b 1 :c]) nil?) => true
+      "Removes top-level to-many refs"
+      (-> (nsh/remove-entity* state [:d 1]) :top-tm) => [[:b 1]]
+      "Removes table-nested to-many refs"
+      (-> (nsh/remove-entity* state [:d 1]) (get-in [:b 1 :c]) nil?) => true
+      "Ignores denormalized data"
+      (-> (nsh/remove-entity* state [:b 1]) (get-in [:denorm :level-1 :level-2])) => denorm-data)))
+
+;============================================================================
+
+(specification "remove-edge*"
+  (component "Simple to-one edge removal"
+    (let [state {:a [:b 1]
```

Comment:
Would `:a [:b 1]` here be considered as an edge? 🤔

Reply:
Yes. That is a root-level prop, and there is a chance that it could be an opaque value (in which case it would not be), but if it were accessed via a query with a join then db->tree would follow it like an ident.  The intention in the spec is to imply the latter.

Reply:
Rename it to `:top-level-edge` if that helps clarity